cpp/src/cylon/table.cpp

```diff
@@ -427,16 +497,298 @@ Status Sort(const std::shared_ptr<Table> &table, const std::vector<int32_t> &sor
   return Table::FromArrowTable(ctx, sorted_table, out);
 }
 
-Status DistributedSort(const std::shared_ptr<Table> &table,
-                       int sort_column,
-                       std::shared_ptr<Table> &output,
-                       bool ascending,
-                       SortOptions sort_options) {
-  return DistributedSort(table, std::vector<int>{sort_column}, output, std::vector<bool>{ascending},
-                         sort_options);
+Status SampleTableUniform(const std::shared_ptr<Table> &local_sorted, 
+                          int num_samples,
+                          std::shared_ptr<Table>& sample_result,
+                          const std::shared_ptr<CylonContext>& ctx) {
+  auto pool = cylon::ToArrowPool(ctx);
+
+  if(local_sorted->Rows() == 0 || num_samples == 0) {
+    std::shared_ptr<arrow::Table> output;
+    RETURN_CYLON_STATUS_IF_ARROW_FAILED(util::CreateEmptyTable(local_sorted->get_table()->schema(),
+                               &output, pool));
+    sample_result = std::make_shared<Table>(ctx, output);
+    return Status::OK();
+  }
+
+  float step = local_sorted->Rows() / (num_samples + 1.0);
+  float acc = step;
+  arrow::Int64Builder filter(pool);
+  RETURN_CYLON_STATUS_IF_ARROW_FAILED(filter.Reserve(num_samples));
+
+  for(int i = 0; i < num_samples; i++) {
+    filter.Append(acc);
+    acc += step;
+  }
+
+  CYLON_ASSIGN_OR_RAISE(auto take_arr, filter.Finish());
+  CYLON_ASSIGN_OR_RAISE(auto take_res, (arrow::compute::Take(local_sorted->get_table(), take_arr)));
+  sample_result = std::make_shared<Table>(ctx, take_res.table());
+
+  return Status::OK();
 }
 
-Status DistributedSort(const std::shared_ptr<Table> &table,
+template<typename T>
+static int CompareRows(const std::vector<std::unique_ptr<T>>& comparators,
+                      int64_t idx_a, int64_t idx_b) { // TODO: priority & direction
+  int sz = comparators.size();
+  if(std::is_same<T, cylon::DualArrayIndexComparator>::value) {
+    idx_b |= (int64_t)1 << 63;
+  }
+  for(int i = 0; i < sz; i++) {
+    int result = comparators[i]->compare(idx_a, idx_b);
+    if(result == 0) continue;
+    return result;
+  }
+  return 0;
+}
+
+Status MergeSortedTable(const std::vector<std::shared_ptr<Table>>& tables,
+                        const std::vector<int>& sort_columns,
+                        const std::vector<bool>& sort_orders,
+                        std::shared_ptr<Table>& out) {
+  std::shared_ptr<Table> concatenated;
+  std::vector<int64_t> table_indices(tables.size()), table_end_indices(tables.size());
+  int acc = 0;
+  for(int i = 0; i < table_indices.size(); i++) {
+    table_indices[i] = acc;
+    acc += tables[i]->Rows();
+    table_end_indices[i] = acc;
+  }
```

Comment:
> By looking at the values of the vectors here, I think you can use a single `tables.size() + 1` vector for this job! then the i'th `table_end_indices` would be `table_indices[i+1]`

I intentionally used 2 tables, because the `table_indices` is used to mark the next element to be merged in this table; values in `table_indices` will be changed and compared to value in `table_end_indices`.

Reply:
I felt like the end result is the same if we use prefix sum.